Makefile

```diff
@@ -12,138 +12,258 @@
-image-windows:
-	docker buildx build --no-cache --build-arg LDFLAGS=$(LDFLAGS) -t $(IMAGE_TAG) -f docker/windows.Dockerfile --platform="windows/amd64" --output "type=docker,push=false" .
-clean:
-	-rm -rf _output
-setup: clean
-	@echo "Setup..."
-	$Q go env
-	$(MAKE) install-helm
-
-ifndef HAS_GOLANGCI
-	curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.30.0
-endif
 
-.PHONY: mod
-mod:
-	@go mod tidy
+.PHONY: image-scan
+image-scan: $(TRIVY)
+	# show all vulnerabilities
+	$(TRIVY) --severity MEDIUM,HIGH,CRITICAL $(IMAGE_TAG)
+	# show vulnerabilities that have been fixed
+	$(TRIVY) --exit-code 1 --ignore-unfixed --severity MEDIUM,HIGH,CRITICAL $(IMAGE_TAG)
 
-KIND_VERSION ?= 0.8.1
-KUBERNETES_VERSION ?= 1.18.2
-VAULT_VERSION ?= 1.4.2
+## --------------------------------------
+## Tooling Binaries
+## --------------------------------------
 
-# USED for windows CI tests
-.PHONY: e2e-test
-e2e-test: e2e-bootstrap
-	make e2e-azure
+$(CONTROLLER_GEN): $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go ## Build controller-gen from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/controller-gen sigs.k8s.io/controller-tools/cmd/controller-gen
 
-.PHONY: e2e-bootstrap
-e2e-bootstrap: install-helm
-	apt-get update && apt-get install bats && apt-get install gettext-base -y
-	# Download and install Vault
-	vault -v | grep -q v$(VAULT_VERSION) || (curl -LO https://releases.hashicorp.com/vault/$(VAULT_VERSION)/vault_$(VAULT_VERSION)_linux_amd64.zip && unzip vault_$(VAULT_VERSION)_linux_amd64.zip && chmod +x vault && mv vault /usr/local/bin/)
-	# Download and install Azure CLI
+$(GOLANGCI_LINT): ## Build golangci-lint from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
+
+$(KUSTOMIZE): ## Build kustomize from tools folder.
+	cd $(TOOLS_MOD_DIR) && \
+		go build -tags=tools -o $(TOOLS_BIN_DIR)/kustomize sigs.k8s.io/kustomize/kustomize/v3
+
+## --------------------------------------
+## Testing Binaries
+## --------------------------------------
+
+$(HELM): ## Install helm3 if not present
+	helm version --short | grep -q v3 || (curl https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3 | bash)
+
+$(KIND): ## Download and install kind
+	kind --version | grep -q $(KIND_VERSION) || (curl -L https://github.com/kubernetes-sigs/kind/releases/download/v$(KIND_VERSION)/kind-linux-amd64 --output kind && chmod +x kind && mv kind /usr/local/bin/)
+
+$(AZURE_CLI): ## Download and install azure cli
 	curl -sL https://aka.ms/InstallAzureCLIDeb | bash
-ifndef TEST_WINDOWS
+
+$(BATS): ## Install bats for running the tests
+	bats --version | grep -q $(BATS_VERSION) || (apt-get -o Acquire::Retries=30 update && apt-get -o Acquire::Retries=30 install -y bats)
```

Comment:
these are different binaries/repos/versions. looks like you are checking for v1.2.1 which is [bats-core](https://github.com/bats-core/bats-core) but I believe apt installed one is [bats](https://github.com/sstephenson/bats) which is the unmaintained older version

Reply:
Updated it!